### DISCUSSION TITLE
Idea to update / fill only missing values of a model  ($model->updateMissing(...))

### DISCUSSION THREAD

**Discussion by LTKort** [laravel/framework#45230] [Ideas]
Hi,

I was wondering if there are any options regarding updating a model, but only fill the fields of that exists on the model and that do not have a existing value. 

So for example I have user model that has these fields;
| id | first_name | last_name | email                |  
|----|------------|-----------|----------------------|
| 1  |            | Kort      | lars@marshmallow.dev |  

And I have an array like; 
```php
 $form_data = [
    'first_name' => 'Lars',
    'last_name' => 'Ananas',
    'phone_number' => '1234567890',
];
```

My goal is to only update the first_name field, since that is the only one empty.

If I call the `$user->update($form_data)` function with the data, it will overwrite the last_name & give an error on phone_number.
Which I don't want, but I can not find any function for this in the docs. 

So if there is no easy solution, I prose a function like `$user->updateMissing($form_data)`. 
Which will check if the attributes exists in the model & updates only the values of the empty keys. 

```php
/**
 * Update only the missing values of the model in the database.
 *
 * @param  array  $attributes
 * @param  array  $options
 * @return bool
 */
public function updateMissing(array $attributes = [], array $options = []
{
    if (!$this->exists) {
        return false;
    }

    $attributes = collect($attributes)
        ->filter(fn ($value, $key) => Arr::has($this->attributesToArray(), $key)) // Removes non-existing keys
        ->reject(fn ($value, $key) => filled($this->$key)) // Removes keys that are already filled
        ->toArray();

    if (empty($attributes)) {
        return false;
    }

    return $this->fill($attributes)->save($options);
}
```

This function works for me, optional is an extra parameter / function to call it quietly. 
Or should this be like an optional parameter in the update function?

Is the above function something I could propose as a PR? 
If so, are there someone that can help me with the tests I should write?
If not, why not? 

Thanks for your help!
;-)

**Comment by henzeb** [comment]
I'd say also add a `fillMissing`, without the save. That way you don't have to clutter the method with a saveQuietly

**Comment by WendellAdriel** [comment]
This seems a nice addition!
If you’re going to work on that I can help you with the tests if you want!

**Reply by LTKort** [reply]
Awesome, should I add a test to the [EloquentUpdateTest](https://github.com/laravel/framework/blob/d23de5f9af1bd253a9d8b823cec9bf37987facf2/tests/Integration/Database/EloquentUpdateTest.php)? 
A function that uses the updateMissing, asserts if a column is filled that was empty and to assert if a column is not overwritten that wasn't empty?

**Reply by LTKort** [reply]
@WendellAdriel 
I created 2 branches; 
 - [Only the updateMissing function](https://github.com/LTKort/laravel-framework/tree/feature/update-missing) with a test 
 - [Added fillMissing](https://github.com/LTKort/laravel-framework/tree/feature/fill-missing) with a separate test which I don't think is in the right location

**Reply by WendellAdriel** [reply]
@LTKort yes, you should add your tests to `EloquentUpdateTest`.
I think that you need to assert at least two things:

- Make sure that the empty column is being updated when you call the `updateMissing`
- Make sure that no other already filled properties are being updated when you call the `updateMissing`

For the `fillMissing` I think the place for the tests would be the `DatabaseEloquentModelTest` and you should have the same assertions for it as for the `updateMissing`.

**Reply by LTKort** [reply]
@WendellAdriel  Thanks for your help.

I added the 'updateMissing() & fillMissing() ' function in the Illuminate\Database\Eloquent\Model.php, but I don't think this is in the correct place. What are the best locations for these functions? Or are they okay where they are?

I added a PR #45272 with 2 test;
For the fill missing in DatabaseEloquentModelTest: 
```php
    public function testFillMissing()
    {
        $model = new EloquentModelStub;
        $model->fillable(['name', 'age']);
        $model->fill(['name' => 'foo', 'age' => null]);
        $model->fillMissing(['name' => 'bar', 'age' => 'baz']);
        $this->assertSame('foo', $model->name);
        $this->assertSame('baz', $model->age);
    }
```
And for updateMissing() in testUpdateMissing:
```php
    public function testUpdateMissing()
    {
        $record = TestUpdateModel1::create([
            'name' => 'LTKort',
            'title' => null,
        ]);

        $record->updateMissing([
            'name' => 'Marshmallow',
            'title' => 'Mr.',
        ]);

        $this->assertSame('LTKort', $record->name);
        $this->assertNotEmpty($record->title);
    }

```

**Reply by WendellAdriel** [reply]
@LTKort 
For the `testUpdateMissing` I would assert the value is `'Mr.'` instead of only asserting is not empty for the `title`. Also, it would be better to use dummy values like `John Doe` or something instead of your username.

For the functions, I think this is the correct place since the `fill` and `update` functions are also in this file!

**Reply by LTKort** [reply]
@WendellAdriel Thanks, I changed the values & changed the `assertNotEmpty` to `assertSame`

**Reply by WendellAdriel** [reply]
I think it's pretty good @LTKort!!!
Amazing work there, a really cool addition!

**Reply by LTKort** [reply]
Thanks for your help & input @WendellAdriel, fingers crossed it will be reviewed (& merged) soon 😄

**Reply by WendellAdriel** [reply]
I hope so @LTKort 🤞🏼

**Comment by stefvanesch** [comment]
Very nice addition!
